### PR TITLE
Handle custom protocols on iOS Safari

### LINKED ISSUE #2268: WhatsApp a href protocol bug
On iOS - Safari : 
the link with "whatsapp" protocol does not open the application WhatsApp:

example:
```html
<a href="whatsapp://send?text=%anytext%"></a>
```

### PR DESCRIPTION
**Pull request by mkhatib**
Fixes #2268

### CODE REVIEW

File: test/functional/test-document-click.js [hunk 1]
```diff
@@ -197,4 +199,46 @@ describe('test-document-click onDocumentElementClick_', () => {
...
+    it('should not do anything on other platforms when embedded', () => {
+      sandbox.stub(platform, 'isIos').returns(true);
+      sandbox.stub(platform, 'isSafari').returns(false);
+      onDocumentElementClick_(evt, viewport);
+      expect(win.top.location.href).to.equal('https://google.com');
+      expect(preventDefaultSpy.callCount).to.equal(0);
+    });
+
+    it('should not do anything if not embedded', () => {
+      win.parent = undefined;
```
Comment (jridgewell) [inline test/functional/test-document-click.js]: You'll need to stub `isIos` and `isSafari` to return true, too.

File: test/functional/test-document-click.js [hunk 2]
```diff
@@ -197,4 +199,46 @@ describe('test-document-click onDocumentElementClick_', () => {
...
+      sandbox.stub(platform, 'isIos').returns(true);
+      sandbox.stub(platform, 'isSafari').returns(true);
+      onDocumentElementClick_(evt, viewport);
+      expect(win.top.location.href).to.equal('whatsapp://send?text=hello');
+      expect(preventDefaultSpy.callCount).to.equal(1);
+    });
+
+    it('should not do anything on other platforms when embedded', () => {
+      sandbox.stub(platform, 'isIos').returns(true);
+      sandbox.stub(platform, 'isSafari').returns(false);
```
Comment (jridgewell) [inline test/functional/test-document-click.js]: This is testing Chrome on iOS, not other platforms. Can you add another test for Android?

File: src/document-click.js [hunk 3]
```diff
@@ -108,6 +113,21 @@ export function onDocumentElementClick_(e, viewport) {
...
+
+  // On Safari iOS, custom protocol links will fail to open apps when the
+  // document is iframed - in order to go around this, we set the top.location
+  // to the custom protocol href.
+  const isSafariIOS = platform.isIos() && platform.isSafari();
+  const isEmbedded = win.parent && win.parent != win;
+  const isCustomProtocolHref = (
+      tgtLoc.protocol != 'http:' && tgtLoc.protocol != 'https:');
+  if (isSafariIOS && isEmbedded && isCustomProtocolHref) {
+    win.top.location.href = target.href;
```
Comment (jridgewell) [inline src/document-click.js]: Is this a security issue? Like if the protocol is `javascript`?

Comment (mkhatib) [inline src/document-click.js]: Isn't `javascript:` protocol invalid in AMP documents and won't be available in the CDN? Thought invalid AMP might still be embedded elsewhere in that case maybe yeah. @cramforce do we need to blacklist some protocols here like `javascript:`?

Comment (dvoytenko) [inline src/document-click.js]: Correct. We can assume that there won't be any insecure protocols here - it's already resolved before this code takes over.

Comment (jridgewell) [inline src/document-click.js]: Excellent, I read this as opening the parent window for some reason and didn't know if this was installed on 3p iframes.

File: src/document-click.js [hunk 4]
```diff
@@ -108,6 +113,21 @@ export function onDocumentElementClick_(e, viewport) {
   const win = doc.defaultView;
 
   const tgtLoc = parseUrl(target.href);
+
+  // On Safari iOS, custom protocol links will fail to open apps when the
+  // document is iframed - in order to go around this, we set the top.location
+  // to the custom protocol href.
+  const isSafariIOS = platform.isIos() && platform.isSafari();
```
Comment (dvoytenko) [inline src/document-click.js]: I'd assume the same failure would be observed on iOS Chrome. Could you test there as well?

File: src/document-click.js (hunk 3 above)
Comment (dvoytenko) [inline src/document-click.js]: Did you try these options?
```
window.open(URL, '_blank');
```
or
```
window.open(URL, '_top');
```

This is the order of my preference if any of these work.

File: src/document-click.js (hunk 4 above)
Comment (mkhatib) [inline src/document-click.js]: It works on iOS Chrome. The problem only seems to happen on iOS Safari.

File: src/document-click.js (hunk 3 above)
Comment (mkhatib) [inline src/document-click.js]: These both work. Though `_blank` seems to be slightly more annoying as it jumps the user to a new tab first and then launches the app. Go with `_blank` anyway?

Comment (mkhatib) [inline src/document-click.js]: Done. Used open _blank.

File: src/document-click.js (hunk 4 above)
Comment (mkhatib) [inline src/document-click.js]: Done.

File: src/document-click.js [hunk 5]
```diff
@@ -108,6 +113,21 @@ export function onDocumentElementClick_(e, viewport) {
...
+  const isEmbedded = win.parent && win.parent != win;
+  const isCustomProtocolHref = (
```
Comment (jridgewell) [inline src/document-click.js]: It'd be easier to test for `isNormalProtocol`, then `!` in the conditional below:

```js
const isNormalProtocol = /^https?:$/.test(tgtLoc.protocol)
if (isSafari && isEmbedded && !normalProtocol) {
  //...
}
```

Comment (mkhatib) [inline src/document-click.js]: Thanks! Done.

**Comment by mkhatib** [comment]
@jridgewell thanks! PTAL :dancer:

File: test/functional/test-document-click.js [hunk 6]
```diff
@@ -197,4 +199,58 @@ describe('test-document-click onDocumentElementClick_', () => {
...
+      sandbox.restore();
+      sandbox = null;
+    });
+
+    it('should set top.location.href on Safari iOS when embedded', () => {
+      sandbox.stub(platform, 'isIos').returns(true);
+      sandbox.stub(platform, 'isSafari').returns(true);
+      onDocumentElementClick_(evt, viewport);
+      expect(win.open.called).to.be.true;
+      expect(win.open.calledWith('whatsapp://send?text=hello')).to.be.true;
```
Comment (dvoytenko) [inline test/functional/test-document-click.js]: Test `_blank` as well.

File: src/document-click.js [hunk 7]
```diff
@@ -108,6 +113,20 @@ export function onDocumentElementClick_(e, viewport) {
...
   const tgtLoc = parseUrl(target.href);
+
+  // On Safari iOS, custom protocol links will fail to open apps when the
+  // document is iframed - in order to go around this, we set the top.location
+  // to the custom protocol href.
+  const isSafariIOS = platform.isIos() && platform.isSafari();
+  const isEmbedded = win.parent && win.parent != win;
+  const isNormalProtocol = /^https?:$/.test(tgtLoc.protocol);
+  if (isSafariIOS && isEmbedded && !isNormalProtocol) {
+    win.open(target.href, '_blank');
```
Comment (dvoytenko) [inline src/document-click.js]: As a bug fix - we should just roll with `_blank`. If experience looks not great, let's file a bug to consider using `_top` instead. That would open up another layer or two of testing - hence the new bug. In the meantime, another question - what target do these URLs arrive now with from `cdn.ampproject.org`?

**Comment by dvoytenko** [comment]
Looks good: just couple of comments/requests.

File: src/document-click.js (hunk 7 above)
Comment (mkhatib) [inline src/document-click.js]: Inside the Google Viewer all links seem to have a `target="_blank"` added to them including these ones. However, this doesn't solve it for iOS Safari (I mean having target=blank on the link).

Comment (dvoytenko) [inline src/document-click.js]: Certainly, that's understood. But it does mean that going with `blank_` is sufficient right now. if we want to change behavior to `top_` - we'd want to do it cdn to start with.

File: test/functional/test-document-click.js (hunk 6 above)
Comment (mkhatib) [inline test/functional/test-document-click.js]: Done.

**Comment by mkhatib** [comment]
Thanks @dvoytenko squashed and addressed the comments.

**Comment by dvoytenko** [comment]
LGTM

**Pull request merged**